compiler/src/dmd/dsymbolsem.d

```diff
@@ -1219,10 +1219,12 @@ private extern(C++) final class DsymbolSemanticVisitor : Visitor
                             bool needctfe = dsym.isDataseg() || (dsym.storage_class & STC.manifest);
                             if (needctfe)
                                 sc = sc.startCTFE();
+                            sc.varDecl = dsym; // https://issues.dlang.org/show_bug.cgi?id=24051
```

Comment:
This is a problem for two reasons:
1. sc.varDecl may have a value in it, but on line 1227 it will reset it to null, i.e. it should save/restore the original value
2. sc may be referenced elsewhere, and altering its state may cause very weird bugs. It's best to push/pop a new sc, which will also neatly solve issue (1).